src/components/Input/Input.tsx

```diff
@@ -77,47 +94,48 @@ export const Input = (
     addonBefore: addonBeforeProp,
   }: InputProps
 ) : ReactElement => {
-  const [val, setVal] = useState({
-    value: value || defaultValue || '',
-    ...getAddonDefaultValue('before', addonBeforeProp),
-    ...getAddonDefaultValue('after', addonAfterProp),
+  const [val, setVal] = useInputValue(value || defaultValue, valuePropName, {
+    before: addonBeforeProp,
+    after: addonAfterProp,
   })
 
   const handleChange = useCallback((event: ChangeEvent<HTMLInputElement>) => {
     const { value: nextValue } = event.target
-    const nextVal = { ...val, value: nextValue }
+    const nextVal = { ...val, [valuePropName]: nextValue }
     setVal(nextVal)
     if (onChange) {
-      onChange(event, nextVal)
-    }
-  }, [onChange, val])
-
-  const renderAddon = useCallback((position: AddonPosition, props: InputAddonProps) => {
-    const { onChange: onChangeAddon, disabled: addonDisabled, ...rest } = props
-
-    const isAddonDisabled = addonDisabled === undefined
-      ? isDisabled
-      : addonDisabled
-
-    const handleChangeAddon = (nextValue: unknown) : void => {
-      const nextVal = { ...val, [position]: nextValue }
-      setVal(nextVal)
-      if (onChangeAddon) {
-        onChangeAddon(nextValue)
-      }
-      if (onChange) {
-        onChange(undefined, nextVal)
-      }
+      onChange(event, addonBeforeProp || addonAfterProp ? nextVal : nextValue)
     }
-
+  }, [addonAfterProp, addonBeforeProp, onChange, valuePropName, setVal, val])
+
+  const renderAddon = useCallback((
+    position: AddonPosition,
+    {
+      value: addonValue,
+      onChange: onChangeAddon,
+      disabled: addonDisabled,
+      name: addonName,
+      ...rest
+    }: InputAddonProps
+  ) => {
     return (
       <InputAddon
-        isDisabled={isAddonDisabled}
-        onChange={handleChangeAddon}
         {...rest}
+        isDisabled={addonDisabled === undefined ? isDisabled : addonDisabled}
+        value={addonValue !== undefined ? addonValue : val[addonName || position]}
+        onChange={(nextValue) => {
+          const nextVal = { ...val, [addonName || position]: nextValue }
+          setVal(nextVal)
+          if (onChange) {
+            onChange(undefined, nextVal)
+          }
+          if (onChangeAddon) {
+            onChangeAddon(nextValue)
+          }
+        }}
       />
     )
-  }, [isDisabled, onChange, val])
+  }, [isDisabled, onChange, setVal, val])
```

Comment:
This renderAddon changes to every key down the user did for the addon input case. I think it loses a little bit of value in memoization

Reply:
this is just for silencing the linter, setVal is actually immutable as it's the setState method from a customHook

Reply:
correct. i've just fixed it